Fedora Pungi mashing including module mashing.

puiterwijk opened the pull request.

igor-raits commented:
btw, does this PR mean that we will switch to pungi for mashing RPMs so we can proceed with rust packaging in fedora?

dustymabe commented:
btw can we properly get my commits included in this patch set? here is the PR I opened: https://github.com/fedora-infra/bodhi/pull/1821. These show up in this patchset as a single `foo` commit.

dustymabe commented:
we probably should not be embedding pungi configs in bodhi - can we delete the `fedora-atomic.conf` commit?

dustymabe commented:
BTW we don't need to clone comps any more as pungi can now do that for us. The [RFE](https://pagure.io/pungi/issue/5) was resolved in 4.1.18

bowlofeggs commented:
@ignatenkobrain Yes, this patch will make Bodhi stop using mash and use Pungi to make RPM repos, OSTrees, and modules too. Exciting!

@dustymabe I'm actually planning to squash all of these commits down to one once we get this ready to merge, but I will attribute authorship to everyone involved in the commit message. Otherwise Bodhi's commit history will be really crazy (Bodhi has a contribution guideline about how commits should be clean and atomic.)

Conan-Kudo commented:
Woohoo!

dustymabe commented:
> @dustymabe I'm actually planning to squash all of these commits down to one once we get this ready to merge, but I will attribute authorship to everyone involved in the commit message. Otherwise Bodhi's commit history will be really crazy (Bodhi has a contribution guideline about how commits should be clean and atomic.)

i'm pretty sure that's what I tried to do in #1821 - just look at [this](https://github.com/fedora-infra/bodhi/pull/1821/commits/630fb807983486e65cecd9417e553983910e816e) commit message and see how much information is included in it. I'd prefer not to squash everything into one massive commit.

bowlofeggs commented:
On 10/16/2017 03:24 PM, Dusty Mabe wrote:
> i'm pretty sure that's what I tried to do in #1821
> <https://github.com/fedora-infra/bodhi/pull/1821> - just look at this
> <https://github.com/fedora-infra/bodhi/pull/1821/commits/630fb807983486e65cecd9417e553983910e816e>
> commit message and see how much information is included in it. I'd
> prefer not to squash everything into one massive commit.

I'm happy to make the commit message include the information we need,
but I think the one massive commit is the way to go because I will need
to cherry pick this feature to release it and because no commits will
work on their own without the full set (thus, they are not atomic). The
feature needs all of them to work, so it should be one commit.

bowlofeggs commented:
I'll remove the ```fedora-atomic.conf```.

dustymabe commented:
> I'm happy to make the commit message include the information we need, but I think the one massive commit is the way to go because I will need to cherry pick this feature to release it

Seems like this feature is so big we'd want a brand new release for it? 

> and because no commits will work on their own without the full set (thus, they are not atomic). The feature needs all of them to work, so it should be one commit.

Isn't that the point of a merge commit? I'm not saying we have to be super vigilant about breaking things into super small pieces, but concise commits with good messages about a specific diff is definitely more useful than a giant commit.

bowlofeggs commented on bodhi/server/consumers/masher.py:
```diff
@@ -623,56 +687,106 @@ def remove_pending_tags(self):
+                                        # Stderr should also go to pungi.global.log if it starts
+                                        stderr=subprocess.PIPE,
+                                        # We will never have additional input
+                                        stdin=DEVNULL)
```
I know /dev/null is special, but will there be a problem if multiple processess are given the same file handle to it? Should we make a call to ```open()``` for each process rather than opening one at the module-level?

bowlofeggs commented on bodhi/server/consumers/masher.py:
```diff
@@ -623,56 +687,106 @@ def remove_pending_tags(self):
+    def get_previous_compose(self):
+        """Function to return the previous composes' output."""
+        return None
+        return os.path.join(config.get('mash_stage_dir'), self.id)
```
Is the intention to kill this method, or was this ```return None``` just here for debugging?

puiterwijk replied:
This was for debugging.

bowlofeggs commented on bodhi/server/consumers/masher.py:
```diff
@@ -707,27 +821,64 @@ def generate_testing_digest(self):
+                self.log.exception("Repodata sanity check failed!")
+                raise
+
+            # make sure that pungi didn't symlink our packages
```
I was actually never really sure why Bodhi does this symlink check (it did it for mash too) - was there some problem in the past that symlinks caused? Does it happen often? Do you think we definitely need this?

puiterwijk replied:
It can create symlinks if you misconfigure pungi.
The problem is that symlinks become problematic with the syncing to /pub.
So this check is really there to make sure that Pungi wasn't misconfigured and got us a repo we can't sync out.

bowlofeggs commented on bodhi/server/consumers/masher.py:
```diff
@@ -752,10 +903,19 @@ def wait_for_sync(self):
-        mash_path = os.path.join(self.path, self.id)
-        arch = os.listdir(mash_path)[0]
+        ## will need to fix mash_path here
+        mash_path = os.path.join(self.path, 'compose', 'Everything')
```
What do we need to fix about this?

puiterwijk replied:
Nothing anymore, that was fixed.

bowlofeggs commented on bodhi/server/consumers/masher.py:
```diff
@@ -917,53 +1077,6 @@ def sort_by_days_in_testing(self, updates):
-                    msg=dict(tag=tag,
-                             ref=result.get('ref'),
-                             commitid=result.get('commitid')),
-                    force=True)
```
I can't wait to kill fedmsg-atomic-composer (which, humorously, isn't triggered via fedmsgs).

puiterwijk replied:
Well, fedmsg-atomic-composer does have a fedmsg consumer and can be ran through purely fedmsg.... :)

bowlofeggs commented on bodhi/server/consumers/masher.py:
```diff
@@ -999,57 +1112,65 @@ def _get_master_repomd_url(self, arch):
+    def copy_additional_pungi_files(self, pungi_conf_dir, template_env):
+        self.git_clone('https://pagure.io/fedora-comps.git',
+                       'master',
+                       os.path.join(pungi_conf_dir, 'comps'))
```
Dusty had suggested using Pungi to do this - I'm slightly inclined to make that a later commit/release just for reducing churn, but if you think that's easy we can do it now.

dustymabe replied:
yeah that's what I was suggesting with https://github.com/fedora-infra/bodhi/pull/1900#issuecomment-336984945 - this is a large enough change that I would say we should just go ahead and leverage the new functionality in pungi

puiterwijk replied:
That's pretty easy.
When I looked at the documentation for it, I couldn't find a way to make it run "make", which in fedora-comps is required.
But after looking at the code, you can do that, so this entire call nd the "make" call can just be dropped.

bowlofeggs commented on bodhi/server/consumers/masher.py:
```diff
@@ -999,57 +1112,65 @@ def _get_master_repomd_url(self, arch):
+            self.log.error('out: %s', out)
+            self.log.error('Err: %s', err)
+            raise Exception('Unable to make comps')
+        shutil.copy('/etc/bodhi/variants-fedora.xml', pungi_conf_dir)
```
Will variants be specific to the release, or do all Fedora releases have the same variants document?

puiterwijk replied:
Let's make it just a template.

bowlofeggs reviewed (commented):
Thanks everyone, this patch looks pretty good so far! I'm going to start off by rebasing this onto https://github.com/fedora-infra/bodhi/pull/1903 once it's merged, and then I'll make a few minor tweaks of my own, and then I'll get to work writing tests and docs. This is a very exciting patch!

bowlofeggs commented:
On 10/16/2017 03:48 PM, Dusty Mabe wrote:
> Seems like this feature is so big we'd want a brand new release for it?

It is getting its own release (well, almost. I'm going to sneak a
backwards-incompatible CLI change as well, since we're naming this one
Bodhi 3.0.0, but that one doesn't affect the server very much), which is
why I will need to cherry pick it. There are many things on develop that
I don't want to release with this patch, in case we need to back it out.
This patch is going to go in with a FBR so I need it to be an easy
downgrade.

> Isn't that the point of a merge commit? I'm not saying we have to be
> super vigilant about breaking things into super small pieces, but
> concise commits with good messages about a specific diff is definitely
> more useful than a giant commit.

I also avoid merge commits, in order to keep a nice clean git history.
My goal is to make it so that each commit on develop could be used as a
release. This has a number of advantages:

1. I can easily cherry pick commits I want to bring into a release (I
   do this often).
2. Other bodhi deployments can also easily cherry pick the commits they
   want.
3. The git log is short and meaningful, without very many "oops fix
   typo" type commits (I did literally fix a typo today with a commit,
   but it was one that had already gotten into develop ☺).
4. Each feature and bug fix has exactly one commit, so it's easy to know
   you got it all when cherry picking.
5. Writing release notes is very painful when commits are not atomic.

It's important to keep in mind that the commit author isn't necessarily
the author of the patch, in general. While it's true that they are often
the same in Bodhi, there are quite a few cases where they are not, and
that's OK. There have been other commits in Bodhi that worked this way
too. When more than one person collaborates on the same patch, the right
thing to do is to credit them in the commit message, similar to how it
would work if we authored a journal paper together. Bodhi also credits
contributors to a release in the release notes, and so all the patch
authors will also be credited there. The commit author is really just
the person who created the commit.

By the way, the requirement for commits to be atomic is documented in
Bodhi's contribution guidelines:

https://bodhi.fedoraproject.org/docs/developer_docs.html#contribution-guidelines

dustymabe commented:
> I also avoid merge commits, in order to keep a nice clean git history. My goal is to make it so that each commit on develop could be used as a release. This has a number of advantages: 

> 1. I can easily cherry pick commits I want to bring into a release (I do this often). 
> 2. Other bodhi deployments can also easily cherry pick the commits they want. 
> 3. The git log is short and meaningful, without very many "oops fix typo" type commits (I did literally fix a typo today with a commit, but it was one that had already gotten into develop ☺). 

I would never propose that we keep "oops fix typo" commits. Anyone that submits a PR should know how to properly rebase/squash commits and get rid of meaningless stuff like that. 

> 4. Each feature and bug fix has exactly one commit, so it's easy to know you got it all when cherry picking.

A large feature in a single commit would probably lose a lot of details about "why". When "changing the way things are" like we are doing in this PR, we change where we look for things and we also delete code in some cases. Without the commit history we lose all of that context. I guess you can try to put it all in a single commit message, but I think most of it will get lost. A merge commit would keep the atomic requirement without losing the history. I don't really like merge commits either, but with the atomic requirement (which you have explained good reasoning for), they would solve the problem. 

> 5. Writing release notes is very painful when commits are not atomic. 

> It's important to keep in mind that the commit author isn't necessarily the author of the patch, in general. While it's true that they are often the same in Bodhi, there are quite a few cases where they are not, and that's OK. There have been other commits in Bodhi that worked this way too. When more than one person collaborates on the same patch, the right thing to do is to credit them in the commit message, similar to how it would work if we authored a journal paper together. Bodhi also credits contributors to a release in the release notes, and so all the patch authors will also be credited there. The commit author is really just the person who created the commit.

I really don't care as much about having my name on commits. I hope I don't seem too vain :). I was more interested in keeping the history and meaningful commit messages that I worked to achieve. 

> By the way, the requirement for commits to be atomic is documented in Bodhi's contribution guidelines: https://bodhi.fedoraproject.org/docs/developer_docs.html#contribution-guidelines

yes I saw that, which is why I suggested a merge commit which would keep the meaningful history and also provide the atomicity required. 

I'll stop beating this horse now. 🏇

bowlofeggs commented:
I've rebased this PR locally, but weirdly GitHub is saying "Add more commits by pushing to the pungi-bodhi-modular branch on puiterwijk/bodhi", while git is saying "ERROR: Permission to puiterwijk/bodhi.git denied to bowlofeggs." I don't know what to make of that.

bowlofeggs commented:
OK, I've rebased the PR and made a few tiny style tweaks so far, now it's time to focus on getting the tests into shape. It looks like there's a lot of work to do:

```
=== 33 failed, 957 passed, 7 pytest-warnings, 1 error in 171.45 seconds ===

Failure. Coverage is below 100.0%.
-------------
Diff Coverage
Diff: develop...HEAD, staged, and unstaged changes
-------------
bodhi/server/consumers/masher.py (16.8%): Missing lines 85-90,196-200,202-207,211,213-215,218,228-230,251-252,254,257,263-264,266-268,270,272,274,276-277,279-283,292-295,392,401,403,445-446,688,690,693,696,699-700,702-705,707-708,710,712-713,715,717,720,725,727-730,734-735,737-738,749,752-758,760,764,767,770-775,778,781-785,820,836,838,842-843,846,849-851,854-856,858,862-864,871-876,878,880-882,896-897,911-916,918,1120,1122-1123,1131,1133,1135-1136,1146,1155,1158-1161,1163-1166,1168,1171-1176,1178
bodhi/server/metadata.py (14.8%): Missing lines 44-47,49-54,56,58,60,62,64-67,102,104,129-130,298
-------------
Total:   223 lines
Missing: 186 lines
Coverage: 16%
-------------
```

It's hard to estimate at this point, but this will definitely take me a few days to sort out.

bowlofeggs commented:
Making progress - I'm down to 27 test failures now!

igor-raits commented:
@bowlofeggs removing tests is always nice =)

bowlofeggs commented:
@ignatenkobrain it sure is an easy way to get them all to pass ☺

bowlofeggs commented:
I may have spoken too soon about the 27 test failures - I hadn't noticed that the ```test_metadata.py``` tests couldn't load due to an import issue so none of them have run yet. I'll work on getting them at least running next, but in the meantime, here's an update on the tests that at least run:

* Down to 12 failures (from 27), not including what failures might happen in the metadata.py tests.
* Up to 54% coverage on the new patch code (from 16%).

bowlofeggs commented:
I've got all the tests running now by fixing that import error, and here's what we're looking at now:

```
=== 18 failed, 976 passed, 7 pytest-warnings in 223.80 seconds ===
Failure. Coverage is below 100.0%.
-------------
Diff Coverage
Diff: develop...HEAD, staged, and unstaged changes
-------------
bodhi/server/consumers/masher.py (60.2%): Missing lines 206-207,268,270,688,690,693,696,754-758,767,770-775,778,781-785,820,836,838,842-843,846,849-851,854-856,858,862-864,871-876,878,880-882,896-897,912,916,1131,1133,1135-1136,1146,1155,1158-1161,1163-1166,1168,1171-1176,1178
bodhi/server/metadata.py (22.2%): Missing lines 44-47,49-54,56,58,60,62,64-67,129-130,298
-------------
Total:   223 lines
Missing: 99 lines
Coverage: 55%
-------------
```

bowlofeggs commented:
Patrick pointed out that this very likely fixes https://github.com/fedora-infra/bodhi/issues/887. When we make our final commit message, we should put "fixes #887" in it. In general, it'd be good to go through the masher bugs and see what other ones we can assume are fixed now too.

bowlofeggs commented:
Thanks to Patrick splitting the work with me, we've made great progress today:

```
=== 6 failed, 984 passed, 7 pytest-warnings in 240.38 seconds ===
Failure. Coverage is below 100.0%.
-------------
Diff Coverage
Diff: develop...HEAD, staged, and unstaged changes
-------------
bodhi/server/consumers/masher.py (67.2%): Missing lines 207-208,269,271,689,745-749,765-766,811,827,829,833-834,837,840-842,845-847,849,853-855,862-867,869,871-873,887-888,903,907,1122,1124,1126-1127,1137,1146,1149-1152,1154-1157,1159,1162-1167,1169
bodhi/server/metadata.py (97.5%): Missing lines 133
-------------
Total:   235 lines
Missing: 65 lines
Coverage: 72%
-------------
```

bowlofeggs commented:
All tests now pass! Next we need to write some new ones to get the coverage up to 100%.

```
=== 991 passed, 7 pytest-warnings in 241.62 seconds ===
Failure. Coverage is below 100.0%.
-------------
Diff Coverage
Diff: develop...HEAD, staged, and unstaged changes
-------------
bodhi/server/consumers/masher.py (75.4%): Missing lines 207-208,269,271,689,745-749,765-766,811,834,847,863-867,869,871-873,903,907,1122,1124,1126-1127,1137,1146,1149-1152,1154-1157,1159,1162-1167,1169
bodhi/server/metadata.py (100%)
bodhi/server/util.py (100%)
-------------
Total:   236 lines
Missing: 48 lines
Coverage: 79%
-------------
```

puiterwijk commented:
```
=== 1000 passed, 18 warnings in 235.12 seconds ===
Diff Coverage
Diff: origin/develop...HEAD, staged and unstaged changes
-------------
bodhi/server/consumers/masher.py (88.3%): Missing lines 814,1126,1128,1130-1131,1141,1150,1153-1156,1158-1161,1163,1166-1171,1173
bodhi/server/metadata.py (100%)
bodhi/server/util.py (100%)
-------------
Total:   237 lines
Missing: 23 lines
Coverage: 90%
```

puiterwijk commented:
```
==== 1001 passed, 20 warnings in 240.63 seconds ===
Diff Coverage
Diff: origin/develop...HEAD, staged and unstaged changes
-------------
bodhi/server/consumers/masher.py (100%)
bodhi/server/metadata.py (100%)
bodhi/server/util.py (100%)
-------------
Total:   237 lines
Missing: 0 lines
Coverage: 100%
```

bowlofeggs commented:
Wow, very nice work Patrick!

bowlofeggs commented on bodhi/server/consumers/masher.py:
```diff
@@ -150,24 +172,66 @@ def consume(self, msg):
+                    # catches the same assumption breakage. This check here is extra for the
+                    # time when someone adds multitype updates and forgets to update this.
+                    raise ValueError('Builds of multiple types found in %s'
+                                     % title)
```
Thanks for the comment, and agreed on the no cover.

bowlofeggs reviewed (approved):
@puiterwijk I think I am good with the current patch set (though of course, I would want to squash it down before merging). I guess we'll wait and see what Jenkies thinks about the latest patch, and I'd like you to review the commits I made tonight too. If you and Jenkies both approve the PR I think we will be good to merge!

bowlofeggs commented:
I did a brief survey of masher bugs, and here is a list of fixes lines I think we can include in the final commit message when we squash all this:

```
fixes #653
fixes #887
fixes #909
fixes #1182
fixes #1219
fixes #1330
```

For nice clickyness, here's those issues: #653, #887, #909, #1182, #1219, #1330.

bowlofeggs commented:
Jenkies looks real backed up, but I ran the container tests on my laptop and they all passed here. So it'll probably pass… eventually…

bowlofeggs commented:
OK this is now squashed down to one commit. @puiterwijk, do you approve of this PR as it stands?

bowlofeggs commented:
I received an IRC approval to merge from @puiterwijk:

```
<fm-apps> github.pull_request.synchronize -- bowlofeggs synchronized pull request #1900 on fedora-infra/bodhi https://github.com/fedora-infra/bodhi/pull/1900
<fm-apps> github.issue.comment -- bowlofeggs commented on issue #1900 on fedora-infra/bodhi https://github.com/fedora-infra/bodhi/pull/1900#issuecomment-337911959
<bowlofeggs> puiterwijk: it is sqashed now - do you also approve the PR?
<bowlofeggs> if so, we can merge and i can start working on a beta
<puiterwijk> bowlofeggs: Yep. Make it so
<bowlofeggs> fantastic
```

bowlofeggs committed:
Use Pungi instead of Mash to generate repositories.

This commit is a significant change in Bodhi, as it alters
the repository layout that Bodhi generates when mashing. It
also enables Bodhi to create repositories of a new content
type, modules. Many other new features of Pungi are now
available to us as well, such as rich dependencies.

fixes #653
fixes #887
fixes #909
fixes #1182
fixes #1219
fixes #1330

This patch was written collectively by several authors:

Kushal Das
Dusty Mabe
Adam Miller
Patrick Uiterwijk
Randy Barlow

Signed-off-by: Randy Barlow <randy@electronsweatshop.com>
Signed-off-by: Adam Miller <maxamillion@fedoraproject.org>
Signed-off-by: Patrick Uiterwijk <patrick@puiterwijk.org>

bowlofeggs commented:
This has been cherry-picked to the 3.0 branch as 87267cc76cc7e5f50da0545e66b8d82658453d91.
